semcore/ellipsis/src/Ellipsis.tsx

```diff
@@ -0,0 +1,124 @@
+import Tooltip from '@semcore/tooltip';
+import { Flex } from '@semcore/flex-box';
+import { useResizeObserver } from './useResizeObserver';
+
+import style from './style/ellipsis.shadow.css';
+import reactToText from '@semcore/utils/lib/reactToText';
+import getOriginChildren from '@semcore/utils/lib/getOriginChildren';
+
+type AsProps = {
+  maxline?: number;
+  trim?: 'end' | 'middle';
+  tooltip?: string;
+  styles?: React.CSSProperties;
+  resizeObserver?: { width: number };
+  // eslint-disable-next-line ssr-friendly/no-dom-globals-in-module-scope
+  containerRef?: RefObject<HTMLElement | null>;
+};
+
+type AsPropsMiddle = {
+  text: string;
+  tooltip?: string;
+  styles?: React.CSSProperties;
+  resizeObserver?: { width: number };
+  // eslint-disable-next-line ssr-friendly/no-dom-globals-in-module-scope
+  containerRef?: RefObject<HTMLElement | null>;
+};
+
+const getSymbolWidth = (fontSize: string) => {
+  const dateSpan = document.createElement('span');
+  dateSpan.setAttribute('style', `fontSize: ${fontSize}px`);
+  dateSpan.innerHTML = 'a';
+  document.body.appendChild(dateSpan);
+  const rect = dateSpan.getBoundingClientRect();
+  dateSpan.remove();
+  return rect.width;
+};
+
+const getFontSize = (element: HTMLElement | null) =>
+  element ? window.getComputedStyle(element, null).getPropertyValue('font-size') : '14';
+
+const getSymbolAmount = (blockWidth: number, symbolWidth: number) =>
+  Math.round(blockWidth / symbolWidth);
+
+class RootEllipsis extends Component<AsProps> {
+  static displayName = 'Ellipsis';
+  static style = style;
+  static defaultProps: AsProps = {
+    trim: 'end',
+    tooltip: 'visible',
+  };
+
+  render() {
+    const SEllipsis = this.Root;
+    const SContainer = Flex;
+    const { styles, Children, maxline, tooltip, trim, resizeObserver, containerRef } = this.asProps;
+    const text = reactToText(getOriginChildren(Children));
+
+    if (trim === 'middle') {
+      return sstyled(styles)(
+        <EllipsisMiddle
+          text={text}
+          styles={styles}
+          tooltip={tooltip}
+          resizeObserver={resizeObserver}
+          containerRef={containerRef}
+        />,
+      );
+    }
+    if (tooltip) {
+      return sstyled(styles)(
+        <SContainer interaction="hover" title={text} tag={Tooltip}>
+          <SEllipsis use:maxline={maxline} render="div" tag="div">
+            <Children />
+          </SEllipsis>
+        </SContainer>,
+      );
+    }
+    return sstyled(styles)(
+      <SEllipsis use:maxline={maxline} render="div" tag="div">
+        <Children />
+      </SEllipsis>,
+    );
+  }
+}
+
+const EllipsisMiddle: React.FC<AsPropsMiddle> = (props) => {
+  const { styles, text, tooltip, resizeObserver, containerRef } = props;
+  const resizeElement = useRef(null);
+  const blockWidth = useResizeObserver(resizeElement, resizeObserver).width;
+  const element = containerRef?.current ?? resizeElement.current;
```

Comment:
You are now allowed to use Refs in such way cause they will always be `null` on first render and require second render to render with correct sizes. Actually I can't get what causes second render and why it even work